### ISSUE TITLE
Invalid function

### ISSUE THREAD

**Issue by abhirocks550** [ParsePlatform/parse-server#764]
#### Environment Setup
I have used Bitnami image for Parse Setup on AWS. There I have cloud code in main.js file. I am able to call hello function using curl request. There i have created one more function as follows:

Parse.Cloud.define('demo',function(req,res){

       res.success('Hello From AWS');
});

#### Steps to reproduce
I tried calling using following curl 
1) 
    curl -X POST \
   -H "X-Parse-Application-Id: myappid" \
   -H "X-Parse-REST-API-Key: ab**" \
   -H "Content-Type: application/json" \
   -d '{}' \
    http://ec2-52-77-221-66.ap-southeast-1.compute.amazonaws.com/parse/functions/demo

2)
 
  curl -X POST \
  -H "X-Parse-Application-Id: myappid" \
  -H "Content-Type: application/json" \
  -d '{}' \
  http://ec2-52-77-221-66.ap-southeast-1.compute.amazonaws.com/parse/functions/demo

#### Logs/Trace

I get following response.
{"code":141,"error":"Invalid function."}

Please Can someone help ?

**Comment by peterdotjs** [comment]
Did you implement the `hello` function? Your repro steps show `hello` while your environment show you've created the `demo` function.

**Comment by abhirocks550** [comment]
Thanks @peterdotjs , sorry that was by mistaked i edited it now.
Yes hello function is there which i am able to call successfuly, but my question is that why i am not able to call demo? While calling demo i am getting {"code":141,"error":"Invalid function."} even when the function is already defined in main.js file.

**Comment by gfosco** [comment]
It sounds like your cloud code file is not being interpreted... You could add a top level `console.log('hello from cloud code');` and verify that it's not being output when the server loads...  Check your ParseServer initialization, and make sure you are providing an absolute path to main.js in the cloud option, as seen here: https://github.com/ParsePlatform/parse-server-example/blob/master/index.js

**Comment by abhirocks550** [comment]
Thanks for reply @gfosco . I am using https://aws.amazon.com/marketplace/pp/B01BLQ17TO this AMI image where all required setup is already done. I think In this parse server.js is running in background so where can i see console.log printed? 

I have added following in server.js where ParserServer initialization is done.

 cloud:process.env.CLOUD_CODE_MAIN || '/home/bitnami/apps/parse/htdocs/cloud '.

After adding this is it necessary to restart ParseServer after adding this changes?

**Comment by gfosco** [comment]
I can't really support the AMI image, don't have any involvement on that... Can you clone the parse-server-example repository and run this locally to test?

**Comment by abhirocks550** [comment]
Thanks @gfosco. Yes i will clone this and will try to run locally.

**Comment by gfosco** [comment]
Clone the example repo, https://github.com/parseplatform/parse-server-example run npm install, then look at index.js for usage.  It's possible to run from this repo directly, but I think for most apps the best approach is using it as a module as in the example.

**Comment by Jerland2** [comment]
Was there ever a solution to this, I am having the same exact problem, except running on Heroku?

**Comment by renatocrobledo** [comment]
Same here! trying with heroku and locally without success.

beforeSave and afterSave are not working too... there's something missing?

**Comment by barnaclejive** [comment]
I have noticed that main.js does not update when the application is redeployed. I am using elastic beanstalk and deploying using 'eb deploy'. I figured this out by checking timestamps of the main.js on the eb server at /var/app/current/cloud/main.js 
Am I missing a step?

**Comment by felipebelluco** [comment]
Any solutions? I'm stuck at this point too!

**Comment by flovilmart** [comment]
It seems that your code is not deployed properly... Double check elastic beanstalk documentation. We can't help much as this is provider specific.

**Comment by felipebelluco** [comment]
I'm actually using Heroku. For curiosity, I tried to call the function hello, that default function that comes with main.js, and oddly it works, and the other functions I actually want call doesn't. I don't have the function hello in my main.js anymore. How's that can happen?

Get Outlook for iOS

		_____________________________

**Comment by jenlai1345** [comment]
I had the same issue too.
Turn out at some point in time, I did a "git init" in my parse-server-example directory to initialize a repository....and then I made changes to to my /cloud/main.js.
Pushed the codes to amazon EB via "eb deploy" ---> it never deployed the updated main.js 

fix: did a "git commit -a" to commit code changes to my local parse-server-example, and then "eb deploy" to push it
The reason you were getting "invalid function" was because on Heroku/AWS it still contains the default/un-updated main.js  

good luck!

**Comment by felipebelluco** [comment]
Thanks mate, I already had fixed it. In my case I just added the host url at the first line of my main.js and it worked.

**Comment by danielchangsoojones** [comment]
yes, @jenlai1345 is right. The new code has not been deployed yet to the actual server, so the server still looks toward the past code. So, you just have to deploy it. If you are using Heroku, this is  how I did it. You literally just have to click a button. 
1) go to the heroku online dashboard and click on your app.
2) got to the "Deploy" tab
3) scroll down to the "Manual Deploy" section and hit the "Deploy Branch" button. Make sure your changes have been merged into the branch you choose to deploy with. I had mine changes merged into the master branch. This button basically tells the Heroku server to update to the new code you have written. 
![screen shot 2016-10-09 at 4 22 09 pm](https://cloud.githubusercontent.com/assets/16820348/19223398/9ae56d96-8e3c-11e6-915f-d70365b665f0.png)
